File: simulator/backend/routes/engagement.ts
import express from 'express';
import { engagementService } from '../services/engagement-service.ts';

const router = express.Router();

// Middleware to check authentication from Bearer token
async function requireAuth(req: express.Request, res: express.Response, next: express.NextFunction) {
  try {
    const token = req.headers.authorization?.replace('Bearer ', '');
    if (!token) {
      return res.status(401).json({ success: false, error: 'Authentication required' });
    }

    const { sessionRepository } = await import('../db/repositories/session-repository.ts');
    const session = await sessionRepository.findByToken(token);

    if (!session) {
      return res.status(401).json({ success: false, error: 'Invalid or expired session' });
    }

    const { userService } = await import('../services/user-service.ts');
    const user = await userService.getUserById(session.user_id);
    if (!user) {
      return res.status(401).json({ success: false, error: 'User not found' });
    }

    (req as any).user = user;
    next();
  } catch (error) {
    console.error('Auth error:', error);
    return res.status(401).json({ success: false, error: 'Authentication failed' });
  }
}

// POST /api/engagement - Track engagement event
router.post('/', requireAuth, async (req: express.Request, res: express.Response) => {
  try {
    const user = (req as any).user;
    const { eventType, simulationId, eventData, sessionId } = req.body;

    // Validate required fields
    if (!eventType) {
      return res.status(400).json({
        success: false,
        error: 'Missing required field: eventType'
      });
    }

    // Validate event type
    const validEventTypes = ['login', 'simulation_start', 'simulation_complete', 'page_view', 'interaction'];
    if (!validEventTypes.includes(eventType)) {
      return res.status(400).json({
        success: false,
        error: `Invalid event type. Must be one of: ${validEventTypes.join(', ')}`
      });
    }

    // Track engagement
    const engagement = await engagementService.trackEngagement({
      userId: user.id,
      eventType,
      simulationId: simulationId || null,
      eventData: eventData || null,
      sessionId: sessionId || null,
    });

    if (!engagement) {
      return res.status(400).json({ success: false, error: 'Failed to track engagement' });
    }

    return res.status(201).json({ success: true, engagement });
  } catch (error) {
    console.error('Track engagement error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to track engagement'
    });
  }
});

// GET /api/engagement/stats - Get engagement statistics
router.get('/stats', requireAuth, async (req: express.Request, res: express.Response) => {
  try {
    const user = (req as any).user;
    const userId = req.query.userId as string;

    // Non-admins can only view their own stats
    const targetUserId = (user.role === 'super_admin' || user.role === 'company_admin') && userId
      ? userId
      : user.id;

    const stats = await engagementService.getEngagementStats(targetUserId);

    return res.json({ success: true, stats });
  } catch (error) {
    console.error('Get engagement stats error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to get engagement stats'
    });
  }
});

// GET /api/engagement/history - Get engagement history
router.get('/history', requireAuth, async (req: express.Request, res: express.Response) => {
  try {
    const user = (req as any).user;
    const userId = req.query.userId as string;
    const startDate = req.query.startDate as string;
    const endDate = req.query.endDate as string;
    const limit = req.query.limit ? parseInt(req.query.limit as string) : 50;

    // Non-admins can only view their own history
    const targetUserId = (user.role === 'super_admin' || user.role === 'company_admin') && userId
      ? userId
      : user.id;

    const history = await engagementService.getEngagementHistory(
      targetUserId,
      startDate ? new Date(startDate) : undefined,
      endDate ? new Date(endDate) : undefined,
      limit
    );

    return res.json({ success: true, history });
  } catch (error) {
    console.error('Get engagement history error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to get engagement history'
    });
  }
});

export default router;

File: backend-python/app/routers/engagement.py
"""
Engagement Router
Engagement tracking endpoints
"""

import logging
from datetime import datetime
from typing import Annotated, Any

from fastapi import APIRouter, HTTPException, status, Depends, Body

from app.services.engagement_service import engagement_service
from app.middleware.auth import get_current_user
from app.models.user import UserData

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/engagement", tags=["Engagement"])


@router.post("", status_code=status.HTTP_201_CREATED)
async def track_engagement(
    user: Annotated[UserData, Depends(get_current_user)],
    eventType: str = Body(...),
    simulationId: str | None = Body(None),
    eventData: dict[str, Any] | None = Body(None),
    sessionId: str | None = Body(None),
):
    """Track engagement event."""
    try:
        # Validate required fields
        if not eventType:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Missing required field: eventType",
            )

        # Validate event type
        valid_event_types = ["login", "simulation_start", "simulation_complete", "page_view", "interaction"]
        if eventType not in valid_event_types:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid event type. Must be one of: {', '.join(valid_event_types)}",
            )

        engagement = await engagement_service.track_engagement({
            "user_id": user.id,
            "event_type": eventType,
            "simulation_id": simulationId,
            "event_data": eventData,
            "session_id": sessionId,
        })

        if not engagement:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Failed to track engagement")

        return {"success": True, "engagement": engagement.model_dump()}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Track engagement error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/stats")
async def get_engagement_stats(
    user: Annotated[UserData, Depends(get_current_user)],
    userId: str | None = None,
):
    """Get engagement statistics."""
    try:
        # Non-admins can only view their own stats
        target_user_id = userId if (user.role in ["super_admin", "company_admin"] and userId) else user.id

        stats = await engagement_service.get_engagement_stats(target_user_id)

        return {"success": True, "stats": stats}
    except Exception as e:
        logger.error(f"Get engagement stats error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/history")
async def get_engagement_history(
    user: Annotated[UserData, Depends(get_current_user)],
    userId: str | None = None,
    startDate: str | None = None,
    endDate: str | None = None,
    limit: int = 50,
):
    """Get engagement history."""
    try:
        # Non-admins can only view their own history
        target_user_id = userId if (user.role in ["super_admin", "company_admin"] and userId) else user.id

        # Parse dates if provided
        start_dt = datetime.fromisoformat(startDate) if startDate else None
        end_dt = datetime.fromisoformat(endDate) if endDate else None

        history = await engagement_service.get_engagement_history(
            target_user_id,
            start_dt,
            end_dt,
            limit,
        )

        return {"success": True, "history": [h.model_dump() for h in history]}
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=f"Invalid date format: {e}")
    except Exception as e:
        logger.error(f"Get engagement history error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))
